Modules/_abc.c

```diff
@@ -0,0 +1,826 @@
+"Internal ABC helper to reset both caches of a given class.\n\
+\n\
+Should be only used by refleak.py");
+
+static PyObject *
+_reset_caches(PyObject *m, PyObject *args)
+{
+    PyObject *self;
+    _abc_data *impl;
+    if (!PyArg_UnpackTuple(args, "_reset_caches", 1, 1, &self)) {
+        return NULL;
+    }
+    impl = _get_impl(self);
+    if (impl == NULL) {
+        return NULL;
+    }
+    if (PySet_Clear(impl->_abc_cache) < 0) {
+        Py_DECREF(impl);
+        return NULL;
+    }
+    /* also the second cache */
+    if (PySet_Clear(impl->_abc_negative_cache) < 0) {
+        Py_DECREF(impl);
+        return NULL;
+    }
+    Py_DECREF(impl);
+    Py_RETURN_NONE;
+}
+
+PyDoc_STRVAR(_get_dump_doc,
+"Internal ABC helper for cache and registry debugging.\n\
+\n\
+Return shallow copies of registry, of both caches, and\n\
+negative cache version. Don't call this function directly,\n\
+instead use ABC._dump_registry() for a nice repr.");
+
+static PyObject *
+_get_dump(PyObject *m, PyObject *args)
+{
+    PyObject *self, *registry, *cache, *negative_cache;
+    _abc_data *impl;
+    if (!PyArg_UnpackTuple(args, "_get_dump", 1, 1, &self)) {
+        return NULL;
+    }
+    impl = _get_impl(self);
+    if (impl == NULL) {
+        return NULL;
+    }
+    registry = PySet_New(impl->_abc_registry);
+    if (registry == NULL) {
+        Py_DECREF(impl);
+        return NULL;
+    }
+    cache = PySet_New(impl->_abc_cache);
+    if (cache == NULL) {
+        Py_DECREF(impl);
+        Py_DECREF(registry);
+        return NULL;
+    }
+    negative_cache = PySet_New(impl->_abc_negative_cache);
+    if (negative_cache == NULL) {
+        Py_DECREF(impl);
+        Py_DECREF(registry);
+        Py_DECREF(cache);
+        return NULL;
+    }
+    Py_INCREF(impl->_abc_negative_cache_version); /* PyTuple_Packdoesn't do this. */
+    PyObject *res = PyTuple_Pack(4,
+            registry, cache, negative_cache, impl->_abc_negative_cache_version);
+    Py_DECREF(registry);
+    Py_DECREF(cache);
+    Py_DECREF(negative_cache);
+    Py_DECREF(impl->_abc_negative_cache_version);
+    Py_DECREF(impl);
+    return res;
+}
+
+// Compute set of abstract method names.
+static int
+compute_abstract_methods(PyObject *self)
+{
+    int ret = -1;
+    PyObject *abstracts = PyFrozenSet_New(NULL);
+    if (abstracts == NULL) {
+        return -1;
+    }
+
+    PyObject *ns = NULL, *items = NULL, *bases = NULL;  // Py_XDECREF()ed on error.
+
+    /* Stage 1: direct abstract methods. */
+    ns = _PyObject_GetAttrId(self, &PyId___dict__);
+    if (!ns) {
+        goto error;
+    }
+
+    /* TODO: Fast path for exact dicts with PyDict_Next */
+    items = PyMapping_Items(ns);
+    if (!items) {
+        goto error;
+    }
+    assert(PyList_Check(items));
+    for (Py_ssize_t pos = 0; pos < PyList_GET_SIZE(items); pos++) {
+        PyObject *it = PySequence_Fast(
+                PyList_GET_ITEM(items, pos),
+                "items() returned non-sequence item");
```

Comment:
non-sequence iterables still work here; should be "items returned non-iterable item"